Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp

```diff
@@ -2707,6 +2708,270 @@ Optional<Ratio> Parser::parse_ratio(TokenStream<ComponentValue>& tokens)
     return Ratio { static_cast<float>(first_number.token().number_value()) };
 }
 
+// https://www.w3.org/TR/css-syntax-3/#urange-syntax
+Optional<UnicodeRange> Parser::parse_unicode_range(TokenStream<ComponentValue>& tokens)
```

Comment:
I think this would be easier to read if it were broken up into a couple of functions.

It seems the first section is sort of lexing the tokens to validate the syntax / find the end position, then the second section is actually parsing those tokens into a range. The first section is tracking a `matched_grammar` boolean all the way through, and having to check `if (!matched_grammar)` the whole way feels a bit awkward.

So e.g. if we break this up, the `u <dimension-token> '?'*` lexing could look be reduced to something like:

```c++
Optional<UnicodeRange> Parser::parse_unicode_range(TokenStream<ComponentValue>& tokens)
{
    //  u <dimension-token> '?'*
    if (second_token.is(Token::Type::Dimension)) {
        while (is_question_mark(tokens.peek_token()))
            tokens.next_token();
        if (is_ending_token(tokens.peek_token()))
            return create_unicode_range();
        tokens.rewind_to_position(at_second_token);
    }
}

Optional<UnicodeRange> Parser::create_unicode_range()
{
    /* This would be everything starting from the following comment: */
    // NOTE: Now that we've matched the grammar, we can start following the spec's algorithm.
}

Reply:
That is an excellent point! Thanks. :smile: